File: services/30-ws-go/internal/auth/middleware.go
package auth

// WebSocket auth — validates auth on the HTTP upgrade request.
//
// WebSocket clients cannot set custom headers in the browser WebSocket API.
// Two auth paths are supported:
//
//   1. Authorization: Bearer <JWT> in the upgrade headers (server clients).
//   2. ?token=<JWT> query parameter in the WebSocket URL (browser clients).
//
// On valid auth: returns *Claims. Caller attaches to connection state.
// On missing or invalid auth: returns nil. Caller closes with 1008 Policy Violation.

import (
	"net/http"
	"strings"
)

// AuthenticateUpgrade extracts and verifies auth from a WebSocket upgrade request.
// Returns the Claims on success.
// Returns nil on auth failure.
func AuthenticateUpgrade(r *http.Request) *Claims {
	// --- Attempt 1: Authorization header (server-side clients) ---
	if authHeader := r.Header.Get("Authorization"); authHeader != "" {
		parts := strings.SplitN(authHeader, " ", 2)
		if len(parts) == 2 && strings.EqualFold(parts[0], "bearer") {
			claims, err := VerifyToken(parts[1])
			if err == nil {
				return claims
			}
		}
	}

	// --- Attempt 2: ?token= query parameter (browser clients) ---
	if tokenParam := r.URL.Query().Get("token"); tokenParam != "" {
		claims, err := VerifyToken(tokenParam)
		if err == nil {
			return claims
		}
	}

	return nil
}

File: services/30-ws-rust/src/auth/middleware.rs
use axum::{
    extract::FromRequestParts,
    http::{request::Parts, StatusCode},
};
use sqlx::PgPool;

use crate::auth::jwt::verify_token;

/// Authenticated user attached to WebSocket upgrade handlers via Axum extractors.
#[derive(Debug, Clone)]
pub struct AuthUser {
    pub id: String,
    pub email: String,
    pub name: String,
}

/// Axum extractor — validates auth on the HTTP upgrade request.
///
/// WebSocket clients cannot set custom headers in browser WebSocket API.
/// Two auth paths are supported:
///
/// 1. `Authorization: Bearer <JWT>` header — for server-side / native clients.
/// 2. `?token=<JWT>` query parameter — for browser WebSocket clients.
///
/// When neither header nor token param is present: 401.
/// When provided but invalid: 401.
impl FromRequestParts<PgPool> for AuthUser {
    type Rejection = StatusCode;

    async fn from_request_parts(
        parts: &mut Parts,
        _pool: &PgPool,
    ) -> Result<Self, Self::Rejection> {
        // --- attempt 1: Bearer JWT in header (server clients) ---
        if let Some(bearer) = parts
            .headers
            .get("Authorization")
            .and_then(|v| v.to_str().ok())
            .and_then(|v| v.strip_prefix("Bearer "))
        {
            let claims = verify_token(bearer).map_err(|_| StatusCode::UNAUTHORIZED)?;
            return Ok(AuthUser {
                id: claims.sub,
                email: claims.email,
                name: claims.name,
            });
        }

        // --- attempt 2: ?token= query parameter (browser WebSocket clients) ---
        if let Some(query) = parts.uri.query() {
            for (key, value) in url::form_urlencoded::parse(query.as_bytes()) {
                if key == "token" {
                    let claims = verify_token(&value).map_err(|_| StatusCode::UNAUTHORIZED)?;
                    return Ok(AuthUser {
                        id: claims.sub,
                        email: claims.email,
                        name: claims.name,
                    });
                }
            }
        }

        Err(StatusCode::UNAUTHORIZED)
    }
}
